lib/luacrypto.c

```diff
@@ -0,0 +1,489 @@
+#include <linux/kernel.h>
+#include <linux/module.h>
+#include <linux/crypto.h>
+#include <crypto/hash.h>
+#include <crypto/aead.h>
+#include <linux/err.h>
+#include <linux/scatterlist.h>
+#include <linux/slab.h>
+
+#include <lua.h>
+#include <lualib.h>
+#include <lauxlib.h>
+#include <lunatik.h>
+
+// Generic hash: hash a string with a given algorithm
+static int luacrypto_hash(lua_State *L)
+{
+	size_t datalen;
+	const char *alg = luaL_checkstring(L, 1);
+	const char *data = luaL_checklstring(L, 2, &datalen);
+
+	struct crypto_shash *tfm;
+	struct shash_desc *desc;
+	int digestsize, ret;
+	u8 *digest;
+
+	tfm = crypto_alloc_shash(alg, 0, 0);
+	if (IS_ERR(tfm))
+		return luaL_error(L, "failed to alloc shash tfm");
+
+	digestsize = crypto_shash_digestsize(tfm);
+	digest = kmalloc(digestsize, GFP_ATOMIC);
+	if (!digest)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc digest buffer");
+	}
+
+	desc = kmalloc(sizeof(*desc) + crypto_shash_descsize(tfm), GFP_ATOMIC);
+	if (!desc)
+	{
+		kfree(digest);
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc desc");
+	}
+	desc->tfm = tfm;
+
+	ret = crypto_shash_digest(desc, data, datalen, digest);
+	crypto_free_shash(tfm);
+	kfree(desc);
+
+	if (ret)
+	{
+		kfree(digest);
+		return luaL_error(L, "digest failed (%d)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)digest, digestsize);
+	kfree(digest);
+	return 1;
+}
+
+static int luacrypto_hmac(lua_State *L)
+{
+	size_t keylen, datalen;
+	const char *alg = luaL_checkstring(L, 1);
+	const char *key = luaL_checklstring(L, 2, &keylen);
+	const char *data = luaL_checklstring(L, 3, &datalen);
+
+	char hmac_alg[CRYPTO_MAX_ALG_NAME];
+	snprintf(hmac_alg, sizeof(hmac_alg), "hmac(%s)", alg);
+
+	struct crypto_shash *tfm = crypto_alloc_shash(hmac_alg, 0, 0);
+	if (IS_ERR(tfm))
+		return luaL_error(L, "failed to alloc hmac tfm");
+
+	int ret = crypto_shash_setkey(tfm, key, keylen);
+	if (ret)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to set hmac key (%d)", ret);
+	}
+
+	int digestsize = crypto_shash_digestsize(tfm);
+	u8 *digest = kmalloc(digestsize, GFP_ATOMIC);
+	if (!digest)
+	{
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc digest buffer");
+	}
+
+	struct shash_desc *desc = kmalloc(sizeof(*desc) + crypto_shash_descsize(tfm), GFP_ATOMIC);
+	if (!desc)
+	{
+		kfree(digest);
+		crypto_free_shash(tfm);
+		return luaL_error(L, "failed to alloc desc");
+	}
+	desc->tfm = tfm;
+
+	ret = crypto_shash_digest(desc, data, datalen, digest);
+	crypto_free_shash(tfm);
+	kfree(desc);
+
+	if (ret)
+	{
+		kfree(digest);
+		return luaL_error(L, "hmac failed (%d)", ret);
+	}
+
+	lua_pushlstring(L, (const char *)digest, digestsize);
+	kfree(digest);
+	return 1;
+}
+
+ /*
+	* The crypto object stored as the private data of the lunatik object.
+	* This object wraps a crypto_aead transform and can be extended to support
+	* multiple algorithms and modes.
+	*/
+struct lua_crypto {
+	struct crypto_aead *tfm;
```

Comment:
if you are only using this nested struct, you don't need to create the external one.. you private type can be `struct crypto_aead`.. take a look in luasocket, we do exactly that.. if such struct must be allocated by the kernel API itself, you can use the pointer flag in the class..